### DISCUSSION TITLE
How do you get the graphql lsp to work?

### DISCUSSION THREAD

**Discussion by RobertBrunhage** [mason-org/mason.nvim#1023] [Q&A]
I've installed it through mason and also added a `.graphqlrc` file that points to the schema as well as all the graphql files.

When checking `LspInfo` graphql is attached but I can't bring up the lsp completions and it doesn't show me any lsp if I misspell anything.

I would assume that it would use the schema to know what to provide me but maybe I am missing something?

**Comment by johnsonjo4531** [comment]
Hey I just got mine working sometime after seeing your thread here. I thought I'd share my answer as it seems there aren't that many people out there that have solved this and written their configs online. I was in a certain situation like you.

I needed to follow this advice

> You need both graphql-language-service-cli and graphql-language-service-server installed globally
> - From [bushblade](https://neovim.discourse.group/u/bushblade) in [this comment](https://neovim.discourse.group/t/nvim-lspconfig-cant-get-graphql-lsp-working/2688/15) on the neovim discourse.

**EDIT:** I actually found out I did not need the graphql-language-service-server and it ended up giving an error staying in my lazy-vim config, so I have removed it below.

For me I'm using LazyVim which uses lazy.nvim plugin manager. To solve the above issue. I did it in my lazyvim configs like so: 

First I added those two dependencies and ensured they were installed when installing mason.nvim:
```lua
  {
    "williamboman/mason.nvim",
    opts = {
      ensure_installed = {
        -- ...elided others
        "graphql-language-service-cli", -- required for graphql-lsp
      },
    },
  },
```
Add graphql to nvim-lspconfig
```lua
  {
    "neovim/nvim-lspconfig",
    opts = {
      servers = {
        -- ...elided other servers
        "graphql", -- required for graphql-lsp
      },
    },
  },
```

Bonus: Add highlighting through nvim-treesitter

```lua
  {
    "nvim-treesitter/nvim-treesitter",
    opts = {
      ensure_installed = {
        -- ...elided other configs
        "graphql",
      },
    },
  },
```

**Reply by rv-mtoohig** [reply]
Awesome, thanks for putting this here.

**Comment by edmondop** [comment]
How would I use a separate / local graphql server installation? I have found a bug in the LSP that prevents the LSP to work correctly

**Comment by marcusgajozo** [comment]
lazy.lua

```lua
-- GraphQL Language Server (lê o graphql.config.cjs automaticamente)
-- Mesmo motor da extensão do VS Code: autocomplete / validação / hover.
-- Instalação automática via npm (inclui o `graphql` para não dar exit 1).
do
  local lsp_dir = vim.env.HOME .. "/.local/share/graphql-lsp"
  local bin = lsp_dir .. "/node_modules/.bin/graphql-lsp"

  vim.lsp.config("graphql", {
    cmd = { bin, "server", "-m", "stream" },
    filetypes = { "graphql", "typescript", "typescriptreact", "javascript", "javascriptreact" },
    root_markers = {
      "graphql.config.cjs", "graphql.config.js", "graphql.config.ts", "graphql.config.json",
      ".graphqlrc", ".graphqlrc.cjs", ".graphqlrc.json", "package.json",
    },
  })

  local function enable()
    vim.lsp.enable("graphql")
  end

  if vim.uv.fs_stat(bin) then
    enable() -- já instalado: só habilita
  else
    vim.notify("[graphql] Instalando GraphQL Language Server via npm...", vim.log.levels.INFO)
    vim.system(
      { "sh", "-c", "mkdir -p " .. lsp_dir .. " && cd " .. lsp_dir .. " && npm install graphql-language-service-cli graphql@16" },
      { text = true },
      function(res)
        if res.code == 0 and vim.uv.fs_stat(bin) then
          vim.schedule(enable)
          vim.notify("[graphql] Language Server instalado com sucesso.", vim.log.levels.INFO)
        else
          vim.notify("[graphql] Falha ao instalar o GraphQL LSP:\n" .. (res.stderr or ""), vim.log.levels.ERROR)
        end
      end
    )
  end
end

-- garante que arquivos .gql abram com o filetype correto (graphql)
vim.filetype.add({ extension = { gql = "graphql" } })
```

plugins/graphql.lua

```lua

return {
  -- syntax highlighting para .gql / .graphql e queries em .ts
  {
    "nvim-treesitter/nvim-treesitter",
    opts = {
      ensure_installed = { "graphql" },
    },
  },
}

```
